File: python/graph/dominator-lengauer-tarjan-simple.py
def dfs_preorder(N, G, s):
    order = [-1]*N
    order[s] = 0
    vs = [0]*N
    vs[0] = s
    prt = [-1]*N
    stk = [s]
    it = [0]*N

    cur = 0
    while stk:
        v = stk[-1]
        while it[v] < len(G[v]):
            w = G[v][it[v]]; it[v] += 1
            if order[w] == -1:
                prt[w] = v
                stk.append(w)

                order[w] = cur = cur+1
                vs[cur] = w
                break
        else:
            stk.pop()
    return vs, order, prt

def dominator(N, G, s):
    vs, semi, prt = dfs_preorder(N, G, s)
    doms = [-1]*N

    RG = [[] for i in range(N)]
    for v in range(N):
        for w in G[v]:
            RG[w].append(v)
    bucket = [[] for i in range(N)]

    *anc, = range(N)
    *label, = range(N)

    def s_link(v, w):
        anc[w] = v
    def s_eval(v):
        if v == anc[v]:
            return v
        compress(v, anc[v])
        return label[v]
    def compress(v, w):
        if w == anc[w]:
            return w
        anc[v] = u = compress(w, anc[w])
        if semi[label[w]] < semi[label[v]]:
            label[v] = label[w]
        return u

    for w in reversed(vs[1:]):
        for v in bucket[w]:
            u = s_eval(v)
            if semi[u] < semi[v]:
                doms[v] = u
            else:
                doms[v] = w
        for v in RG[w]:
            su = semi[s_eval(v)]
            if su < semi[w]:
                semi[w] = su
        if vs[semi[w]] != prt[w]:
            bucket[vs[semi[w]]].append(w)
        else:
            doms[w] = prt[w]
        s_link(prt[w], w)
    for v in bucket[s]:
        doms[v] = s
    for w in vs[1:]:
        if doms[w] != vs[semi[w]]:
            doms[w] = doms[doms[w]]
    doms[s] = s
    return doms

File: cpp/graph/dominator-lengauer-tarjan-simple.cpp
#include<vector>
using namespace std;

#define N 100003

int n;
vector<int> g[N];

vector<int> vs;
int order[N];
int prt[N];
int doms[N], semi[N];
int anc[N], label[N];
vector<int> rg[N], bucket[N];

inline void link(int v, int w) {
  anc[w] = v;
}

int compress(int v, int w) {
  if(w == anc[w]) {
    return w;
  }

  anc[v] = compress(w, anc[w]);
  if(semi[label[w]] < semi[label[v]]) {
    label[v] = label[w];
  }
  return anc[v];
}

inline int eval(int v) {
  if(v == anc[v]) {
    return v;
  }

  compress(v, anc[v]);
  return label[v];
}

void dfs(int v) {
  order[v] = vs.size();
  vs.push_back(v);
  for(int w : g[v]) {
    if(order[w] != -1) continue;
    prt[w] = v;
    dfs(w);
  }
}

void dominator(int s) {
  for(int i=0; i<n; ++i) order[i] = -1, prt[i] = -1;
  vs.reserve(n);

  dfs(s);

  for(int i=0; i<n; ++i) semi[i] = order[i], doms[i] = -1;
  for(int i=0; i<n; ++i) anc[i] = label[i] = i;

  for(int v=0; v<n; ++v) {
    for(int w : g[v]) {
      rg[w].push_back(v);
    }
  }

  for(int i=n-1; i>0; --i) {
    int w = vs[i];
    for(int v : bucket[w]) {
      int u = eval(v);
      doms[v] = (semi[u] < semi[v] ? u : w);
    }
    for(int v : rg[w]) {
      int u = eval(v);
      if(semi[u] < semi[w]) {
        semi[w] = semi[u];
      }
    }
    if(vs[semi[w]] != prt[w]) {
      bucket[vs[semi[w]]].push_back(w);
    } else {
      doms[w] = prt[w];
    }
    link(prt[w], w);
  }
  for(int v : bucket[s]) doms[v] = s;

  for(int i=1; i<n; ++i) {
    int w = vs[i];
    if(doms[w] != vs[semi[w]]) {
      doms[w] = doms[doms[w]];
    }
  }
  doms[s] = s;
}